### PR TITLE
[MXNET-563] Refactor R optimizers to fix memory leak

### PR DESCRIPTION
**Pull request by jeremiedb**
Fix R memory leakage through refactor of the optimizers. 
Given the mutatable NDArray isn't supported in R, optimizers have been ported as symbolic update, an executor being created for each weight. 

! Only optimizers update symbols have been used for now, so only SGD, rmsprop and Adam are now supported. Will need to reimplement the manual update for Adagrad and Adadelta. 

Memory is now kept at low level even for very large networks and embeddings. 
Tested on CPU and single GPU, not multiple GPUs.

### CODE REVIEW

**Comment by jeremiedb** [comment]
Provides fix for #10721 #10928

**Comment by thirdwing** [comment]
@hetong007 Please take a look at this.

**Comment by jeremiedb** [comment]
@anirudhacharya Sure, I'll add tests. 
Would be great if you could jump in as well. I was expecting to add the missing Adagrad and Adadelta optimizers within a week in order to match existing functionnalities as soon as possible. Would you be disposed looking at the non-mutatble NDArrays which was actually the root cause leading to refactor optimizers into symbolic execution? Thanks!

**Comment by anirudhacharya** [comment]
@jeremiedb sure! also could you please mail me at [anirudhkrec at gmail]. I would like to sync with you about improvements to the R binding.

**Comment by jeremiedb** [comment]
@hetong007 With Adadelta and Adagrad, the same functionnalities as now supported (and non-centered rmsprop has been added within rmsprop). Tests to be added.

File: R-package/R/optimizer.R [hunk 1]
```diff
@@ -1,32 +1,70 @@
...
                        lr_scheduler = NULL) {
-  # use lr as short for learing rate.
+  
   lr <- learning.rate
-  count       <- 0
-  num_update  <- 0
-
+  count <- 0
+  num_update <- 0
+
```
Comment (hetong007) [inline R-package/R/optimizer.R]: please eliminate trailing whitespaces.

File: R-package/R/optimizer.R [hunk 2]
```diff
@@ -40,105 +78,116 @@ mx.opt.sgd <- function(learning.rate,
...
 
 #' Create an RMSProp optimizer with respective parameters.
 #' Reference: Tieleman T, Hinton G. Lecture 6.5-rmsprop: Divide the gradient by a running average of its recent magnitude[J]. COURSERA: Neural Networks for Machine Learning, 2012, 4(2).
 #' The code follows: http://arxiv.org/pdf/1308.0850v5.pdf Eq(38) - Eq(45) by Alex Graves, 2013.
-#' 
-#' @param learning.rate float, default=0.002
-#'      Step size.
-#' @param gamma1 float, default=0.95
+#'
+#' @param learning.rate float, default=1e-3
```
Comment (hetong007) [inline R-package/R/optimizer.R]: Is there a strong reason to change the default values? It may break other people's code.

File: R-package/R/optimizer.R [hunk 3]
```diff
@@ -392,31 +459,22 @@ mx.opt.create <- function(name, ...) {
...
 #' @export
-mx.opt.get.updater <- function(optimizer, weights) {
-  # This is the list to keep track of internal states of optimzer
-  state.list <- lapply(seq_along(weights), function(i) {
-    if (is.null(weights[[i]])) return(NULL)
-    optimizer$create.state(i, weights[[i]])
+mx.opt.get.updater <- function(optimizer, weights, ctx) {
+  
+  exec_list <- lapply(seq_along(weights), function(i) {
+    if (is.null(weights[[i]])) return(NULL) else
```
Comment (hetong007) [inline R-package/R/optimizer.R]: please format here and below as

```
if (condition) {
   xxx
} else {
  yyy
}
```

File: R-package/R/optimizer.R (hunk 2 above)
Comment (jeremiedb) [reply R-package/R/optimizer.R]: I wanted to align with the Python's package default. I'll revert to existing default if you see if you see more harms from this change.

Comment (hetong007) [reply R-package/R/optimizer.R]: I suggest to keep the default value as in R package. It's not necessary to set default values the same across interfaces, while it may break user's script if we change it, especially for parameters in optimizer.

**Comment by hetong007** [comment]
You may use the help from your editor to eliminate trailing spaces.  Also, please add tests for new optimizers.

**Comment by jeremiedb** [comment]
@hetong007 Tests added and trailing space fixed.

File: R-package/R/optimizer.R [hunk 4]
```diff
@@ -40,105 +78,117 @@ mx.opt.sgd <- function(learning.rate,
...
 #' Reference: Tieleman T, Hinton G. Lecture 6.5-rmsprop: Divide the gradient by a running average of its recent magnitude[J]. COURSERA: Neural Networks for Machine Learning, 2012, 4(2).
 #' The code follows: http://arxiv.org/pdf/1308.0850v5.pdf Eq(38) - Eq(45) by Alex Graves, 2013.
-#' 
+#'
 #' @param learning.rate float, default=0.002
-#'      Step size.
+#'      The initial learning rate.
 #' @param gamma1 float, default=0.95
 #'      decay factor of moving average for gradient, gradient^2.
 #' @param gamm2 float, default=0.9
```
Comment (hetong007) [inline R-package/R/optimizer.R]: `gamm2` -> `gamma2`

File: R-package/R/optimizer.R [hunk 5]
```diff
@@ -159,41 +209,60 @@ mx.opt.rmsprop <- function(learning.rate=0.002,
...
-                        wd=0,
-                        rescale.grad=1,
-                        clip_gradient = NULL,
+mx.opt.adam <- function(learning.rate = 1e-3,
+                        beta1 = 0.9,
+                        beta2 = 0.999,
+                        epsilon = 1e-8,
+                        wd = 0,
+                        rescale.grad = 1,
+                        clip_gradient = -1,
```
Comment (hetong007) [inline R-package/R/optimizer.R]: what is the difference between setting it to `1` and `-1` ?

Comment (jeremiedb) [reply R-package/R/optimizer.R]: When `clip_gradient` is < 0, no clipping is applied.

Comment (hetong007) [reply R-package/R/optimizer.R]: Please add that to the docstring, otherwise people may feel confused without looking at the code.

**Comment by aplikaplik** [comment]
Hi, i have same problem with GPU memory (package mxnet, fuction mx.mlp, just started with that). But as i'm traffic engineer, i have a little problem with orientation in provided solution (i'm use r for long time but not in this detail). Can you pleas be so kind a provide some "Refactor R optimizers for dumies" version of solution? :).

**Comment by jeremiedb** [comment]
@hetong007 Is there any blocking element remaining?

**Comment by jeremiedb** [comment]
@aplikaplik Is your issue specificly about mxnet MLP function or also related to gpuR? As for mx.mlp, this PR should fix memory encountered with both mx.mlp and mx.model.Feedforward.create since they are all relying on the same optimizers update routine. 
Roughly speaking, the idea of the fix was to create an symbolic graph for each parameter to be trained. At each update, the state weights are update through a forward pass on a graph in which the calculation are those associated with each of the optimisation routine (SGD, Adam, ...). At the end, it's the same approach, but using the symbolic representation (mx.symbol) rather than the imperative interface (mx.nd) since the latter had apparent memory leak.

**Comment by aplikaplik** [comment]
@jeremiedb Hi, i referenced gpuR because it could be usefull for @cdeterman. My problem is much simpler, because i don't know how implement this code (optimizer.R) to my R instalation or mxnet package? Im sorry for that elemtal question.

**Comment by onomatet** [comment]
@hetong007 @jeremiedb 
Hi, could you please check if #17207 is related to the changes in this pull request?
If yes, are there any temporary solutions to alter the learning rate of the optimizers in symbolic representation?
